File: interpreted/dart/nanocode.dart
#!/usr/bin/env dart
// nanocode - minimal claude code alternative (Dart)
import 'dart:convert';
import 'dart:io';

final KEY = Platform.environment['ANTHROPIC_API_KEY']!;
final MODEL = Platform.environment['MODEL'] ?? 'claude-sonnet-4-20250514';
const R = '\x1b[0m', B = '\x1b[1m', D = '\x1b[2m', C = '\x1b[36m', G = '\x1b[32m', BL = '\x1b[34m';

String tool(String name, Map input) {
  try {
    switch (name) {
      case 'read': return File(input['path']).readAsLinesSync().asMap().entries.map((e) => '${e.key + 1}| ${e.value}').join('\n');
      case 'write': File(input['path']).writeAsStringSync(input['content']); return 'ok';
      case 'edit': var t = File(input['path']).readAsStringSync(); if (!t.contains(input['old'])) return 'error: not found';
        File(input['path']).writeAsStringSync(t.replaceFirst(input['old'], input['new'])); return 'ok';
      case 'glob': return Process.runSync('find', ['.', '-name', input['pat']]).stdout.toString().split('\n').take(50).join('\n');
      case 'grep': return Process.runSync('grep', ['-rn', input['pat'], '.']).stdout.toString().split('\n').take(50).join('\n');
      case 'bash': return Process.runSync('sh', ['-c', input['cmd']]).stdout.toString();
    }
  } catch (e) { return 'error: $e'; }
  return 'unknown';
}

final schema = jsonDecode('''[{"name":"read","description":"Read","input_schema":{"type":"object","properties":{"path":{"type":"string"}},"required":["path"]}},
{"name":"write","description":"Write","input_schema":{"type":"object","properties":{"path":{"type":"string"},"content":{"type":"string"}},"required":["path","content"]}},
{"name":"edit","description":"Edit","input_schema":{"type":"object","properties":{"path":{"type":"string"},"old":{"type":"string"},"new":{"type":"string"}},"required":["path","old","new"]}},
{"name":"glob","description":"Find","input_schema":{"type":"object","properties":{"pat":{"type":"string"}},"required":["pat"]}},
{"name":"grep","description":"Search","input_schema":{"type":"object","properties":{"pat":{"type":"string"}},"required":["pat"]}},
{"name":"bash","description":"Run","input_schema":{"type":"object","properties":{"cmd":{"type":"string"}},"required":["cmd"]}}]''');

Future<Map> ask(List messages) async {
  final client = HttpClient();
  final req = await client.postUrl(Uri.parse('https://api.anthropic.com/v1/messages'));
  req.headers.set('Content-Type', 'application/json');
  req.headers.set('anthropic-version', '2023-06-01');
  req.headers.set('x-api-key', KEY);
  req.write(jsonEncode({'model': MODEL, 'max_tokens': 4096, 'system': 'Concise assistant', 'messages': messages, 'tools': schema}));
  final resp = await req.close();
  return jsonDecode(await resp.transform(utf8.decoder).join());
}

void main() async {
  print('${B}nanocode$R | $D$MODEL$R\n');
  var messages = <Map>[];

  while (true) {
    stdout.write('$B${BL}❯$R ');
    final input = stdin.readLineSync()?.trim();
    if (input == null || input == '/q') break;
    if (input.isEmpty) continue;
    if (input == '/c') { messages = []; print('${G}⏺ Cleared$R'); continue; }

    messages.add({'role': 'user', 'content': input});

    while (true) {
      final resp = await ask(messages);
      final content = resp['content'] as List;
      final results = <Map>[];

      for (final block in content) {
        if (block['type'] == 'text') print('\n$C⏺$R ${block['text']}');
        if (block['type'] == 'tool_use') {
          final name = block['name'];
          print('\n$G⏺ $name$R');
          final result = tool(name, block['input']);
          print('  $D⎿ ${result.split('\n').first}$R');
          results.add({'type': 'tool_result', 'tool_use_id': block['id'], 'content': result});
        }
      }

      messages.add({'role': 'assistant', 'content': content});
      if (results.isEmpty) break;
      messages.add({'role': 'user', 'content': results});
    }
    print('');
  }
}

File: interpreted/javascript/node/nanocode.mjs
#!/usr/bin/env node
// nanocode - minimal claude code alternative (Node.js ESM)
import { createInterface } from "readline";
import { execSync } from "child_process";
import { readFileSync, writeFileSync, readdirSync } from "fs";
import { join } from "path";

const KEY = process.env.OPENROUTER_API_KEY || process.env.ANTHROPIC_API_KEY;
const API = process.env.OPENROUTER_API_KEY ? "https://openrouter.ai/api/v1/messages" : "https://api.anthropic.com/v1/messages";
const MODEL = process.env.MODEL || (process.env.OPENROUTER_API_KEY ? "anthropic/claude-opus-4" : "claude-sonnet-4-20250514");
const [R, B, D, C, G] = ["\x1b[0m", "\x1b[1m", "\x1b[2m", "\x1b[36m", "\x1b[32m"];

const walk = (d) => { try { return readdirSync(d,{withFileTypes:1}).flatMap(e => e.isDirectory() ? walk(join(d,e.name)) : [join(d,e.name)]); } catch { return []; }};
const tools = {
  read: (a) => readFileSync(a.path,"utf8").split("\n").slice(a.offset||0,(a.offset||0)+(a.limit||1e9)).map((l,i)=>`${(a.offset||0)+i+1}| ${l}`).join("\n"),
  write: (a) => (writeFileSync(a.path, a.content), "ok"),
  edit: (a) => { const t=readFileSync(a.path,"utf8"); if(!t.includes(a.old)) return "error: not found"; if(!a.all&&t.split(a.old).length>2) return "error: not unique"; writeFileSync(a.path,a.all?t.replaceAll(a.old,a.new):t.replace(a.old,a.new)); return "ok"; },
  glob: (a) => walk(a.path||".").filter(f=>f.includes(a.pat.replace(/\*+/g,""))).slice(0,50).join("\n")||"none",
  grep: (a) => { const re=new RegExp(a.pat); return walk(a.path||".").flatMap(f=>{ try { return readFileSync(f,"utf8").split("\n").map((l,i)=>[f,i+1,l]).filter(([,,l])=>re.test(l)).map(([f,n,l])=>`${f}:${n}:${l}`); } catch { return []; }}).slice(0,50).join("\n")||"none"; },
  bash: (a) => { try { return execSync(a.cmd,{encoding:"utf8",timeout:30000}); } catch(e) { return e.stdout||e.message; }},
};
const schema = [
  {name:"read",description:"Read file",input_schema:{type:"object",properties:{path:{type:"string"},offset:{type:"integer"},limit:{type:"integer"}},required:["path"]}},
  {name:"write",description:"Write file",input_schema:{type:"object",properties:{path:{type:"string"},content:{type:"string"}},required:["path","content"]}},
  {name:"edit",description:"Edit file",input_schema:{type:"object",properties:{path:{type:"string"},old:{type:"string"},new:{type:"string"},all:{type:"boolean"}},required:["path","old","new"]}},
  {name:"glob",description:"Find files",input_schema:{type:"object",properties:{pat:{type:"string"},path:{type:"string"}},required:["pat"]}},
  {name:"grep",description:"Search files",input_schema:{type:"object",properties:{pat:{type:"string"},path:{type:"string"}},required:["pat"]}},
  {name:"bash",description:"Run command",input_schema:{type:"object",properties:{cmd:{type:"string"}},required:["cmd"]}},
];

async function ask(messages) {
  const res = await fetch(API, {
    method: "POST",
    headers: { "Content-Type": "application/json", "anthropic-version": "2023-06-01", ...(process.env.OPENROUTER_API_KEY ? { Authorization: `Bearer ${KEY}` } : { "x-api-key": KEY }) },
    body: JSON.stringify({ model: MODEL, max_tokens: 8192, system: `Concise coding assistant. cwd: ${process.cwd()}`, messages, tools: schema })
  });
  return res.json();
}

console.log(`${B}nanocode${R} | ${D}${MODEL}${R}\n`);
const rl = createInterface({ input: process.stdin, output: process.stdout });
const messages = [];

const prompt = () => rl.question(`${B}\x1b[34m❯${R} `, async (input) => {
  if (!input.trim()) return prompt();
  if (input === "/q") return rl.close();
  if (input === "/c") { messages.length = 0; console.log(`${G}⏺ Cleared${R}`); return prompt(); }
  
  messages.push({ role: "user", content: input });
  while (1) {
    const { content } = await ask(messages);
    const results = [];
    for (const b of content) {
      if (b.type === "text") console.log(`\n${C}⏺${R} ${b.text}`);
      if (b.type === "tool_use") {
        console.log(`\n${G}⏺ ${b.name}${R}(${D}${JSON.stringify(b.input).slice(0,50)}${R})`);
        const r = tools[b.name](b.input);
        console.log(`  ${D}⎿ ${r.split("\n")[0].slice(0,60)}${R}`);
        results.push({ type: "tool_result", tool_use_id: b.id, content: r });
      }
    }
    messages.push({ role: "assistant", content });
    if (!results.length) break;
    messages.push({ role: "user", content: results });
  }
  console.log(); prompt();
});
prompt();
